ffi/pic.lua

```diff
@@ -253,20 +254,39 @@ function Pic.openGIFDocumentFromData(data, size)
 end
 
 function Pic.openPNGDocument(filename)
-    local ok, re = Png.decodeFromFile(filename)
+    local req_n
+    local bbtype
+    if Pic.color then
+        req_n = 3
+    else
+        -- NOTE: LodePNG will NOT do RGB -> Grayscale conversions for us, for design reasons (multiple ways to do it, lossy).
+        --       So we can only *ask* to keep grayscale PNGs as-is, but we might actually be getting fed back a RGB24 one ;).
+        req_n = 1
+    end
+
+    local ok, re = Png.decodeFromFile(filename, req_n)
     if not ok then error(re) end
 
+    if re.ncomp == 1 then bbtype = BB.TYPE_BB8
+    elseif re.ncomp == 2 then bbtype = BB.TYPE_BB8A
+    elseif re.ncomp == 3 then bbtype = BB.TYPE_BBRGB24
+    elseif re.ncomp == 4 then bbtype = BB.TYPE_BBRGB32
+    else
+        error("unsupported number of color components")
+    end
+
     local doc = PicDocument:new{width=re.width, height=re.height}
-    doc.image_bb = BB.new(re.width, re.height, BB.TYPE_BBRGB32, re.data)
-    -- mark buffer for freeing when Blitbuffer is freed:
+    doc.image_bb = BB.new(re.width, re.height, bbtype, re.data)
+    doc.components = re.ncomp
+
+    -- Mark buffer for freeing when Blitbuffer is freed:
     doc.image_bb:setAllocated(1)
-    doc.components = 4
     return doc
 end
 
 function Pic.openJPGDocument(filename)
-    local fh = io.open(filename, "r")
-    assert(fh, "couldn't open file")
+    local fh = io.open(filename, "rb")
+    assert(fh, "couldn't open JPG file")
```

Comment:
These are all pcalled in front, right?

Reply:
Yeah, picdocument's pcalling that (which initially threw me for a loop, as errors get thrown into the void) ;).

Reply:
I tend to think of pcalls without catching the error as a code smell, also see https://github.com/koreader/koreader-base/pull/827